### DISCUSSION TITLE
How to correctly implement a Closure from once_into_js to be called multiple times?

### DISCUSSION THREAD

**Discussion by ronnybremer** [wasm-bindgen/wasm-bindgen#3780] [Q&A] [answered]
### Summary

I am new in the field of wasm-bindgen and am stuck at an issue which I can't seem to implement correctly. Right now, I have a Rust _CallBack<Event>_ function, which should be called from an external JS library. In order to pass the callback to JS, I used the following call which returns a JsValue:
```
let onClick_cb = Closure::once_into_js(move |event: Event| onClick.emit(event));
```

As I understand the documentation, this will basically destroy the Rust Closure object and pass the closure into JS, were it will be automatically destroyed after being called/executed. Works like a charm.

Since I don't want to render my component after each click, I am trying to implement this callback for the components lifecycle. So I tried to use _Closure::new()_ but to no avail. I can only get it to compile inside a _#[wasm_bindgen]_ block, but there I can't use the _Event_ nor the _CallBack_ struct.

```
#[wasm_bindgen]
pub struct Closures {
    onClick_c: Closure<dyn FnMut()>,
}

#[wasm_bindgen]
impl Closures {
    pub fn new(onClick: Callback<()>) -> Closures {
        let onClick_c = Closure::new(|_| onClick.emit(()));
        Closures { onClick_c }
    }
}
```

The idea is to keep the instance of Closures around during the lifecycle of the component and implement a Destroy function to free the Rust Closure. Passing it into JS was then hopefully working like:
```
closures.onClick_c.into_js_value()
```

However, Rust complains with the following error:
```
error[E0277]: the trait bound `yew::Callback<web_sys::Event>: FromWasmAbi` is not satisfied
   --> src/kendoui/button.rs:159:1
    |
159 | #[wasm_bindgen]
    | ^^^^^^^^^^^^^^^ the trait `FromWasmAbi` is not implemented for `yew::Callback<web_sys::Event>`
    |
    = help: the following other types implement trait `FromWasmAbi`:
              bool
              char
              isize
              i8
              i16
              i32
              i64
              usize
            and 155 others
    = note: this error originates in the attribute macro `wasm_bindgen::prelude::__wasm_bindgen_class_marker` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.

```

I read through many articles and how-tos but they either use closures with no arguments or native arguments like u32. I need to pass the Event struct back into the Rust callback somehow.

Any help would be highly appreciated.

### Additional Details

serde = { version = "1.0.195", features = ["derive"] }
serde-wasm-bindgen = "0.6.3"
wasm-bindgen = { version = "0.2.89", features = ["serde-serialize"] }
wasm-bindgen-futures = "0.4.39"

**Comment by daxpedda** [comment, marked as answer]
Are you trying to instantiate `Closures` from JS?
Because if not you should not use `#[wasm_bindgen]` on the `impl` block.

What issue exactly do you have when trying to use `Closure::new()` without `#[wasm_bindgen]`? As far as I could see from your code you just have to use the same parameters as you defined in `Closures::onClick_c`. E.g. `Closure::new(|| ...)` instead of `Closure::new(|_| ...)`.

**Reply by ronnybremer** [reply]
You were absolutely right, thank you for pointing this out. Instantiating the Closure in a separate function (no wasm_bindgen) works like a charm. I can pass in the "event: Event" parameter and it gets processed correctly.
